Streaming file listings through the page cache

chrisvest opened the pull request:
**What?**
Stream all files in target directory that is reachable by page cache. The file handles in the stream let user delete or rename the underlying files.

**Why?**
The page cache needs to be able to list files and do recursive deletion to be able to imitate behaviour of `BranchDataPolicy` (needed to support HA). Blockdevice does not have any concept of directories so streaming only the files and deleting directories emptied by `rename` and `delete` makes more sense than listing content.

chrisvest committed:
First draft of streaming recursive file listings through the page cache

chrisvest committed:
Remove the PageCache.renameFile method because it is superseded by the streamFilesRecursive.

lutovich commented on community/io/src/main/java/org/neo4j/io/fs/DefaultFileSystemAbstraction.java:
```diff
@@ -206,6 +206,12 @@ public long lastModifiedTime( File file )
     }
 
+    @Override
+    public void deleteFileOrThrow( File file ) throws IOException
```
There already exists `boolean deleteFile(File)` method. I'd expect it to use `deleteFileOrThrow()` but it does not and has different implementation (uses `FileUtils` instead of `Files`). This looks quite suspicious. Can we make `deleteFile()` use `deleteFileOrThrow()` and return `true` on success and `false` on exception?

chrisvest replied:
They have slightly different semantics wrt. to deleting files that don't exist, and I didn't want to uproot the use of deleteFile at this point because it has a lot of uses all over the code base. We'd have to examine each to see if they can tolerate the change in semantics.

lutovich replied:
Fair enough, let's not touch `deleteFile()` now.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/FileHandle.java:
```diff
@@ -17,15 +17,19 @@
+
+    File getFile();
+
+    void renameFile( File to, CopyOption... options ) throws IOException;
```
Maybe just `rename()`?

chrisvest replied:
`renameFile` mirrors `FileUtils.renameFile`, but sure we can change it to just `rename`, or to `move` to mirror `java.nio.file.Files.move`.

lutovich replied:
`renameFile()` is just slightly inconsistent with `delete()`. I think both should either have or omit `File` in their names

chrisvest replied:
Fair point.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/FileIsMappedException.java:
```diff
@@ -0,0 +1,59 @@
+
+        private final String message;
+
+        Operation( String message )
```
Should this constructor be public? Class is public and `FileIsMappedException` is fully public as well.

chrisvest replied:
Enum constructors are private. The enum itself is public because it is passed to the `FileIsMappedException` constructor, which is public.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/SingleFilePageSwapperFactory.java:
```diff
@@ -78,6 +84,37 @@ public void renameUnopenedFile( File sourceFile, File targetFile, CopyOption...
+        try
+        {
+            // We grab a snapshot of the file tree to avoid seeing the same file twice or more due to renames.
+            List<File> snapshot = streamFilesRecursiveInner( directory ).collect( toList() );
```
Is eagerness required here? Maybe add a comment explaining why

chrisvest replied:
Yes, we need to snapshot because file renames could otherwise make files appear more than once in the stream.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/SingleFilePageSwapperFactory.java:
```diff
@@ -100,4 +137,54 @@ public long getRequiredBufferAlignment()
         return 1;
     }
+
+    private class WrappingFileHandle implements FileHandle
```
Can this class be top level and get its own unit tests?

chrisvest replied:
It can be, but it's an implementation detail of the `PageSwapperFactory`. If something needs more unit testing it would be the `PSF`.

lutovich commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCache.java:
```diff
@@ -405,16 +407,64 @@ public synchronized void renameFile( File sourceFile, File targetFile, CopyOptio
+        };
+    }
+
+    private void throwIfMapped( File file, FileIsMappedException.Operation operation ) throws IOException
```
I'd vote for `assertNotMapped()` but this is not quite important.

lutovich commented on community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java:
```diff
@@ -4862,4 +4875,434 @@ public void fileSizeMustIncreaseInPageIncriments() throws Exception
+        assertThat( filepaths, containsInAnyOrder( a.getCanonicalPath() ) ); // note that we don't go into 'sub'
+    }
+
+    @Test( expected = NoSuchFileException.class )
```
Can we avoid using `@Test(expected)` and instead use `ExpectedException` rule or `try-catch`? Otherwise it is really hard to understand where exception is expected to be thrown from (especially true for following tests).

chrisvest replied:
Yes, I will do that change once the feature work part has settled down. It'll be a lot of modified lines that should go into a dedicated commit.

lutovich reviewed (changes requested):
@chrisvest review round completed. Made some comments.

chrisvest committed:
Add javadocs to PageCache.streamFilesRecursive and FileHandle

chrisvest committed:
Remove useless FileHandle.getAbsolutePath

chrisvest committed:
Use simpler method nemes in the PageCache.streamFilesRecursive area of the code

chrisvest commented:
@lutovich I have not addressed the following:
- [x] the way exceptions are tested for
- [x] the `PSF` testing
- [x] the javadocs on the new `PSF` APIs

This still needs to be done.

chrisvest committed:
Make FileHandle.delete/rename recursively remove empty parent directories

chrisvest committed:
Use ExpectedException rule instead of @Test(expect = ...) annotation in the page cache tests

This clarifies exactly part of the test should fail.

chrisvest committed:
Also use ExpectedException rule for PageSwapperTest

chrisvest committed:
PageCacheTest cleanups

chrisvest committed:
Add testing for the new file listing APIs in PageSwapperFactory

chrisvest committed:
Add javadocs for PageSwapperFactory.streamFilesRecursive

chrisvest committed:
Remove unused PageSwapperFactory.renameUnopenedFile

chrisvest commented:
@lutovich All missing things are implemented and all of your comments from your earlier review have been addressed.

Please give it another gander.

chrisvest committed:
Extract implementation of SingleFilePageSwapperFactory.streamFilesRecursive into a static helper, available to other PageSwapperFactory implementations

chrisvest commented:
@lutovich The last build has an unrelated test failure. It's just flaky, I think.

chrisvest committed:
Fix a PageSwapperTest

It was testing for file path canonicalisation in the wrong place.
Fixing the test also exposed a bug, which has been fixed as well.
